src/UnitTests/ForEachEnumeratorDispose.cs

```diff
@@ -0,0 +1,434 @@
+        }
+
+        [Fact]
+        public void ForEach_for_unknown_nondisposable_struct_should_not_call_Dispose()
+        {
+            Collection.IsDisposable = false;
+            Collection.ShouldThrowException = false;
+            Collection.UseStruct = true;
+            SourceDest source = new SourceDest { Numbers = new Collection() };
+            Mapper.Map<SourceDest>(source).ShouldNotBeNull();
+        }
+
+        [Fact]
+        public void ForEach_for_unknown_disposable_struct_should_call_Dispose()
+        {
+            Collection.IsDisposable = true;
+            Collection.ShouldThrowException = false;
+            Collection.UseStruct = true;
+            SourceDest source = new SourceDest { Numbers = new Collection() };
+            Mapper.Map<SourceDest>(source).ShouldNotBeNull();
+            source.Numbers.IsDisposed.ShouldBe(true);
+        }
+
+        [Fact]
+        public void ForEach_for_unknown_disposable_struct_should_call_Dispose_despite_of_exception()
+        {
+            Collection.IsDisposable = true;
+            Collection.ShouldThrowException = true;
+            Collection.UseStruct = true;
+            SourceDest source = new SourceDest { Numbers = new Collection() };
+            Should.Throw<Exception>(() => Mapper.Map<SourceDest>(source).ShouldNotBeNull());
+            source.Numbers.IsDisposed.ShouldBe(true);
+        }
+
+        [Fact]
+        public void ForEach_for_known_disposable_class_should_call_Dispose()
+        {
+            GenericCollection.UseStruct = false;
+            GenericCollection.ShouldThrowException = false;
+            SourceDest source = new SourceDest { GenericNumbers = new GenericCollection() };
+            Mapper.Map<SourceDest>(source).ShouldNotBeNull();
+            source.GenericNumbers.IsDisposed.ShouldBe(true);
+        }
+
+        [Fact]
+        public void ForEach_for_known_disposable_class_should_call_Dispose_despite_of_exception()
+        {
+            GenericCollection.UseStruct = false;
+            GenericCollection.ShouldThrowException = true;
+            SourceDest source = new SourceDest { GenericNumbers = new GenericCollection() };
+            Should.Throw<Exception>(() => Mapper.Map<SourceDest>(source).ShouldNotBeNull());
+            source.GenericNumbers.IsDisposed.ShouldBe(true);
+        }
+
+        [Fact]
+        public void ForEach_for_known_disposable_struct_should_call_Dispose()
+        {
+            GenericCollection.UseStruct = true;
+            GenericCollection.ShouldThrowException = false;
+            SourceDest source = new SourceDest { GenericNumbers = new GenericCollection() };
+            Mapper.Map<SourceDest>(source).ShouldNotBeNull();
+            source.GenericNumbers.IsDisposed.ShouldBe(true);
+        }
+
+        [Fact]
+        public void ForEach_for_known_disposable_struct_should_call_Dispose_despite_of_exception()
+        {
+            GenericCollection.UseStruct = true;
```

Comment:
This won't work. Because the struct would be boxed. See [this](https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,569). So you need a proper implementation for struct. But I guess only two tests for struct should do (one IDisposable and one not). I'm not sure we need more.

Reply:
Why would the struct be boxed? Isn't there a direct call of the `Dispose` method?
Do you know any solution for it?

Reply:
Follow that link. The public GetEnumerator returns the struct, that's the optimization I was talking about.

Reply:
I'm sorry, but I still do not understand neither the problem nor the solution.

I can see that the `GetEnumerator` method returns a struct, but where does the boxing happen? Do you want me to check for the non-generic `IEnumerator<T>.GetEnumerator()` method in order to omit the runtime check?

Reply:
The boxing happens in your GetEnumerator because you return an interface, not the concrete type. Again, check how List does it and do the same.

Reply:
I hope, my PR fulfills your requirements.

Reply:
At this point it felt cleaner to start from scratch, which I did in #3158.